release-19.2: fix server hangs on HTTP timeouts

knz opened the pull request:
Backport:
  * 1/1 commits from "*: new `httputil.Client` and methods, & use them" (#42536)
  * 1/1 commits from "cli/interactive_tests: opt out of telemetry by default" (#42537)
  * 1/1 commits from "lint: only run TestRoachLint nightly" (#42595)
  * 1/1 commits from "lint: properly parallelize TestVet/TestVet" (#42605) 

Please see individual PRs for details.

/cc @cockroachdb/release

knz committed:
*: new `httputil.Client` and methods, & use them

Previously, various parts of CockroachDB were using the `http`
package's Get/Post/Head methods/functions directly. This was defective
for two reasons:

- it did not implement a request timeout, which could cause
  requests to hang forever if the server was inoperative.
- it did not obey context cancellation, which would force
  the CockroachDB server to hang forever while trying to
  quit if some HTTP server was blocking.

The last point in particular may cause excessive operator friction
when restarting a server, by forcing a node to be killed and
potentially disrupting operation.

This patch fixes the issue by introducing a new `httputil.Client`
class which implements a customizable timeout, see its constructor
`NewClientWithTimeout`.  The default instance in
`httputil.DefaultClient` uses a 3 second timeout.

It also introduces context-aware variants of the `http`
methods/functions `Get`, `Post` and `Head`.  Note that these variants
do not _actually_ obey context cancellation when the code is compiled
with Go 1.12 or prior. The new behavior only kicks in with Go 1.13 or
later (it requires the new feature `NewRequestWithContext`). However,
the usability problem should still be somewhat alleviated in Go 1.12
builds thanks to the request timeout implemented by the new
`httputil.Client`.

The following components in CockroachDB are then adapted to use the
new interface:

- avro registration for changefeed encoders: timeout + cancellation
- telemetry updates: timeout + cancellation
- binfetcher: timeout + cancellation
- cloudinfo: timeout + cancellation
- docgen: timeout, no cancellation (interactive tool)
- roachprod install: timeout, no cancellation (interactive tool)
- test server: timeout, no cancellation (tests have timeouts already)

Release note (bug fix): CockroachDB will now less likely hang in an
inconvenient/inoperative state if it attempts to access an external
HTTP server that blocks or is overloaded. A possible symptom of
the bug is a node failing to shut down upon `cockroach quit`. This
bug is present since at least version 2.0.

knz committed:
cli/interactive_tests: opt out of telemetry by default

Prior to this patch, the CLI interactive tests were running with
diagnostic reporting (including update server checks, telemetry etc)
enabled.

This was unintentional—having implemented the tests initially, I
always worked under the assumption that the acceptance test runner was
disabling telemetry upfront. I only discovered the opposite was true
recently, seeing tests fail with mysterious errors when the telemetry
server happens to be mis-behaving.

This patch fixes the situation by opting the interactive tests out of
telemetry by default. The two tests that check it (with a sensible
fallback) now opt back in explicitly.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42539)
<!-- Reviewable:end -->

tbg commented:
LGTM but I'll defer to @bdarnell about postponing this for a few days until after 19.2.1.

bdarnell reviewed (approved):
LGTM. This should go into 19.2.2, but the commit for 19.2.1 has already been picked so I think this is ready to go after the master commit is merged.

knz committed:
lint: only run TestRoachLint nightly

TestRoachLint has problematic memory usage which makes
it inadequate to run on every CI, see issue #42594.

This patch moves it to nightly.

Release note: None

knz committed:
lint: properly parallelize TestVet/TestVet

The `shadow` and `printfuncs` testvet invocations are
not amenable to parallelization well.

This patch fixes it.

Release note: None
